nydus-image: refactor unpack/compact cli interface

SToPire opened the pull request:
Since unpack and compact subcommands does not need the entire nydusd configuration file, let's refactor their cli interface and directly take backend configuration file.

Specifically, we introduce `--backend-type`, `--backend-config` and `--backend-config-file` options to specify the backend type and remove `--config` option.

Fixes: #1602

## Relevant Issue (if applicable)
#1602 

## Details
Introduce  `--backend-type`, `--backend-config` and `--backend-config-file` options for `unpack` and `compact` subcommands in nydus-image.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation Update (if none of the other choices apply)

## Checklist

- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.

SToPire commented:
As it may break the existing usage of `nydus-image unpack`, I hope to take more advices for this PR. Then I will add the corresponding tests and update the doc.

imeoer commented on src/bin/nydus-image/main.rs:
```diff
@@ -1778,15 +1797,80 @@ impl Command {
+                    bail!("--backend-config or --backend-config-file must be specified");
+                };
+
+            backend = if backend_type == "localfs" {
```
Maybe refactor `BlobFactory::new_backend` or add a new `BlobFactory::new_backend_from` to create backend from config string, so that we can't miss the handle of the new backend types in future.

imeoer reviewed (commented):
others LGMT, thanks!

SToPire committed:
smoke: fix invalid cleanup issue in main_test.go

The cleanup of new registry is invalid as TestMain() calls os.Exit()
and will not run defer functions. This patch fixes the issue by
doing the cleanup explicitly.

Signed-off-by: Yifan Zhao <zhaoyifan@sjtu.edu.cn>

SToPire committed:
nydus-image: refactor unpack/compact cli interface

Since unpack and compact subcommands does not need the entire nydusd
configuration file, let's refactor their cli interface and directly
take backend configuration file.

Specifically, we introduce `--backend-type`, `--backend-config` and
`--backend-config-file` options to specify the backend type and remove
`--config` option.

Signed-off-by: Yifan Zhao <zhaoyifan@sjtu.edu.cn>

Fixes: #1602

imeoer commented on storage/src/factory.rs:
```diff
@@ -245,6 +248,43 @@ impl BlobFactory {
         }
     }
 
+    pub fn new_backend_from_json(
```
Can we reuse `new_backend` here? we can also reuse the handle of `#[cfg(feature = "backend-xxx")]`.

imeoer replied:
Okay, let's only add the `feature` flag. :)

imeoer commented on smoke/tests/main_test.go:
```diff
@@ -22,10 +22,16 @@ func TestMain(m *testing.M) {
 	code := m.Run()
+
+	if reg != nil {
+		reg.Destroy()
```
Why are we taking it out?

SToPire replied:
As I mentioned in the commit message, the cleanup logic registered with `defer` will not be executed when `os.Exit()` is called.

imeoer commented on src/bin/nydus-image/main.rs:
```diff
@@ -671,17 +671,41 @@ fn prepare_cmd_args(bti_string: &'static str) -> App {
-                        .required(true),
+                    Arg::new("backend-type")
+                        .long("backend-type")
+                        .help("Type of backend [possible values: localdisk, s3, oss, registry, http-proxy]")
```
Maybe provide a list method of backend types in `storage/src/factory.rs` and use it here.

SToPire replied:
Yes, I will try it.

imeoer commented on src/bin/nydus-image/main.rs:
```diff
@@ -703,10 +732,24 @@ fn prepare_cmd_args(bti_string: &'static str) -> App {
+            .arg(
+                Arg::new("backend-type")
+                    .long("backend-type")
+                    .help("Type of backend [possible values: localdisk, s3, oss, registry, http-proxy]")
```
Ditto.

imeoer reviewed (commented):
others LGTM!

imeoer reviewed (approved):
LGTM, thanks!
